File: java/src/test/java/org/libpetri/core/ComposeDirectMembershipTest.java
package org.libpetri.core;

import org.junit.jupiter.api.Test;
import org.libpetri.core.Arc.In;
import org.libpetri.core.Arc.Out;
import org.libpetri.fixtures.SubnetFixtures;

import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Tests for subnet-membership metadata recorded by
 * {@link PetriNet.Builder#compose(SubnetDef)} — per <b>MOD-026</b>.
 *
 * <p>Direct composition records a {@code node-name → subnet-name} mapping on
 * the built net: a node contributed by exactly one subnet maps to that
 * subnet; a place contributed by two or more subnets is a shared rendezvous
 * place and is omitted. The mapping is what cluster-aware DOT export uses to
 * reconstruct {@code subgraph cluster_*} blocks without prefix names.
 */
class ComposeDirectMembershipTest {

    /** A producer whose output place is named {@code pipe} (wires by name). */
    private static SubnetDef<Void> pipeProducer() {
        Place<String> seed = Place.of("seed", String.class);
        Place<String> pipe = Place.of("pipe", String.class);
        return SubnetDef.builder("PipeProducer")
            .place(seed)
            .place(pipe)
            .transition(Transition.builder("emit")
                .inputs(In.one(seed))
                .outputs(Out.place(pipe))
                .build())
            .build();
    }

    /** A consumer whose input place is named {@code pipe} (wires by name). */
    private static SubnetDef<Void> pipeConsumer() {
        Place<String> pipe = Place.of("pipe", String.class);
        Place<String> sink = Place.of("sink", String.class);
        return SubnetDef.builder("PipeConsumer")
            .place(pipe)
            .place(sink)
            .transition(Transition.builder("eat")
                .inputs(In.one(pipe))
                .outputs(Out.place(sink))
                .build())
            .build();
    }

    @Test
    void directCompose_recordsMembershipPerSubnet() {
        var net = PetriNet.builder("Host")
            .compose(pipeProducer())
            .compose(pipeConsumer())
            .build();

        var membership = net.subnetMembership();
        assertEquals("PipeProducer", membership.get("seed"),
            "a single-owner place maps to its subnet");
        assertEquals("PipeProducer", membership.get("emit"),
            "a transition maps to its subnet");
        assertEquals("PipeConsumer", membership.get("sink"));
        assertEquals("PipeConsumer", membership.get("eat"));
    }

    @Test
    void directCompose_sharedPlace_hasNoMembershipEntry() {
        // 'pipe' is contributed by both subnets — a shared rendezvous place
        // with no single owner, so it carries no membership entry (it renders
        // top-level, outside any cluster).
        var net = PetriNet.builder("Host")
            .compose(pipeProducer())
            .compose(pipeConsumer())
            .build();

        assertFalse(net.subnetMembership().containsKey("pipe"),
            "a place contributed by 2+ subnets must have no membership entry");
    }

    @Test
    void directCompose_membershipIsOrderIndependent() {
        var ab = PetriNet.builder("Host")
            .compose(pipeProducer())
            .compose(pipeConsumer())
            .build();
        var ba = PetriNet.builder("Host")
            .compose(pipeConsumer())
            .compose(pipeProducer())
            .build();

        assertEquals(ab.subnetMembership(), ba.subnetMembership(),
            "membership must not depend on compose order");
    }

    @Test
    void flatNet_hasEmptyMembership() {
        Place<String> a = Place.of("a", String.class);
        Place<String> b = Place.of("b", String.class);
        var net = PetriNet.builder("Flat")
            .transition(Transition.builder("move")
                .inputs(In.one(a)).outputs(Out.place(b)).build())
            .build();

        assertTrue(net.subnetMembership().isEmpty(),
            "a hand-written flat net carries no subnet membership");
    }

    @Test
    void instanceCompose_hasEmptyMembership() {
        // compose(Instance) is the prefix-rename path — it records no
        // membership; the exporter clusters those nodes by name prefix.
        var net = PetriNet.builder("Host")
            .compose(SubnetFixtures.producer().instantiate("p1"))
            .build();

        assertTrue(net.subnetMembership().isEmpty(),
            "instance composition records no membership metadata");
    }

    @Test
    void directCompose_thenFuse_dropsNonCanonicalEntry() {
        // Fusion removes the non-canonical place 'b'; its membership entry
        // must be dropped so the map names only places that still exist.
        Place<String> a = Place.of("a", String.class);
        Place<String> b = Place.of("b", String.class);
        var subnet = SubnetDef.builder("AB")
            .transition(Transition.builder("move")
                .inputs(In.one(a))
                .outputs(Out.place(b))
                .build())
            .build();

        var net = PetriNet.builder("Host")
            .compose(subnet)
            .fuse(FusionSet.of("ab", a, b))
            .build();

        assertEquals("AB", net.subnetMembership().get("a"),
            "the canonical fused place keeps its membership entry");
        assertFalse(net.subnetMembership().containsKey("b"),
            "the non-canonical fused place must lose its membership entry");
        assertEquals("AB", net.subnetMembership().get("move"),
            "transitions are unaffected by fusion");
    }

    @Test
    void bindActions_preservesMembership() {
        var net = PetriNet.builder("Host")
            .compose(pipeProducer())
            .compose(pipeConsumer())
            .build();

        var bound = net.bindActions(Map.of("emit", TransitionAction.passthrough()));

        assertEquals(net.subnetMembership(), bound.subnetMembership(),
            "bindActions rebuilds transitions by name — membership must survive");
    }

    @Test
    void directCompose_subnetNameWithSlash_isSanitised() {
        // A subnet name containing '/' must be sanitised so it never triggers
        // spurious nested-cluster splitting in the exporter.
        Place<String> x = Place.of("x", String.class);
        Place<String> y = Place.of("y", String.class);
        var subnet = SubnetDef.builder("group/sub")
            .transition(Transition.builder("step")
                .inputs(In.one(x)).outputs(Out.place(y)).build())
            .build();

        var net = PetriNet.builder("Host").compose(subnet).build();

        assertEquals("group_sub", net.subnetMembership().get("step"),
            "'/' in a subnet name must be replaced with '_'");
    }
}

File: typescript/tests/core/compose-direct-membership.test.ts
import { describe, expect, it } from 'vitest';
import { PetriNet } from '../../src/core/petri-net.js';
import { place } from '../../src/core/place.js';
import { SubnetDef } from '../../src/core/subnet-def.js';
import { Transition } from '../../src/core/transition.js';
import { one } from '../../src/core/in.js';
import { outPlace } from '../../src/core/out.js';
import { FusionSet } from '../../src/core/fusion-set.js';
import { passthrough } from '../../src/core/transition-action.js';
import { producer } from '../fixtures/subnet-fixtures.js';

/**
 * Tests for subnet-membership metadata recorded by
 * `PetriNetBuilder.compose(def)` — per **MOD-026**.
 *
 * Direct composition records a `node-name → subnet-name` mapping on the built
 * net: a node contributed by exactly one subnet maps to that subnet; a place
 * contributed by two or more subnets is a shared rendezvous place and is
 * omitted. The mapping is what cluster-aware DOT export uses to reconstruct
 * `subgraph cluster_*` blocks without prefix names.
 *
 * Mirrors `java/src/test/java/org/libpetri/core/ComposeDirectMembershipTest.java`.
 */
describe('compose(def) subnet-membership metadata — MOD-026', () => {
  /** A producer whose output place is named `pipe` (wires by name). */
  function pipeProducer(): SubnetDef<void> {
    const seed = place<string>('seed');
    const pipe = place<string>('pipe');
    return SubnetDef.builder('PipeProducer')
      .place(seed)
      .place(pipe)
      .transition(Transition.builder('emit').inputs(one(seed)).outputs(outPlace(pipe)).build())
      .build();
  }

  /** A consumer whose input place is named `pipe` (wires by name). */
  function pipeConsumer(): SubnetDef<void> {
    const pipe = place<string>('pipe');
    const sink = place<string>('sink');
    return SubnetDef.builder('PipeConsumer')
      .place(pipe)
      .place(sink)
      .transition(Transition.builder('eat').inputs(one(pipe)).outputs(outPlace(sink)).build())
      .build();
  }

  const sortedEntries = (m: ReadonlyMap<string, string>): [string, string][] =>
    [...m.entries()].sort((a, b) => a[0].localeCompare(b[0]));

  it('directCompose_recordsMembershipPerSubnet', () => {
    const net = PetriNet.builder('Host')
      .compose(pipeProducer())
      .compose(pipeConsumer())
      .build();

    const membership = net.subnetMembership;
    expect(membership.get('seed')).toBe('PipeProducer');
    expect(membership.get('emit')).toBe('PipeProducer');
    expect(membership.get('sink')).toBe('PipeConsumer');
    expect(membership.get('eat')).toBe('PipeConsumer');
  });

  it('directCompose_sharedPlace_hasNoMembershipEntry', () => {
    // 'pipe' is contributed by both subnets — a shared rendezvous place with
    // no single owner, so it carries no membership entry (it renders
    // top-level, outside any cluster).
    const net = PetriNet.builder('Host')
      .compose(pipeProducer())
      .compose(pipeConsumer())
      .build();

    expect(net.subnetMembership.has('pipe')).toBe(false);
  });

  it('directCompose_membershipIsOrderIndependent', () => {
    const ab = PetriNet.builder('Host')
      .compose(pipeProducer())
      .compose(pipeConsumer())
      .build();
    const ba = PetriNet.builder('Host')
      .compose(pipeConsumer())
      .compose(pipeProducer())
      .build();

    expect(sortedEntries(ab.subnetMembership)).toEqual(sortedEntries(ba.subnetMembership));
  });

  it('flatNet_hasEmptyMembership', () => {
    const a = place<string>('a');
    const b = place<string>('b');
    const net = PetriNet.builder('Flat')
      .transition(Transition.builder('move').inputs(one(a)).outputs(outPlace(b)).build())
      .build();

    expect(net.subnetMembership.size).toBe(0);
  });

  it('instanceCompose_hasEmptyMembership', () => {
    // compose(instance) is the prefix-rename path — it records no membership;
    // the exporter clusters those nodes by name prefix.
    const net = PetriNet.builder('Host')
      .compose(producer().instantiate('p1'))
      .build();

    expect(net.subnetMembership.size).toBe(0);
  });

  it('directCompose_thenFuse_dropsNonCanonicalEntry', () => {
    // Fusion removes the non-canonical place 'b'; its membership entry must be
    // dropped so the map names only places that still exist.
    const a = place<string>('a');
    const b = place<string>('b');
    const subnet = SubnetDef.builder('AB')
      .transition(Transition.builder('move').inputs(one(a)).outputs(outPlace(b)).build())
      .build();

    const net = PetriNet.builder('Host')
      .compose(subnet)
      .fuse(FusionSet.of('ab', a, b))
      .build();

    expect(net.subnetMembership.get('a')).toBe('AB');
    expect(net.subnetMembership.has('b')).toBe(false);
    expect(net.subnetMembership.get('move')).toBe('AB');
  });

  it('bindActions_preservesMembership', () => {
    const net = PetriNet.builder('Host')
      .compose(pipeProducer())
      .compose(pipeConsumer())
      .build();

    const bound = net.bindActions(new Map([['emit', passthrough()]]));

    expect(sortedEntries(bound.subnetMembership)).toEqual(sortedEntries(net.subnetMembership));
  });

  it('directCompose_subnetNameWithSlash_isSanitised', () => {
    // A subnet name containing '/' must be sanitised so it never triggers
    // spurious nested-cluster splitting in the exporter.
    const x = place<string>('x');
    const y = place<string>('y');
    const subnet = SubnetDef.builder('group/sub')
      .transition(Transition.builder('step').inputs(one(x)).outputs(outPlace(y)).build())
      .build();

    const net = PetriNet.builder('Host').compose(subnet).build();

    expect(net.subnetMembership.get('step')).toBe('group_sub');
  });
});
